examples/with_winit/src/minimal_pipeline_cache.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2025 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Code related to managing wgpu pipeline caches, especially on Android
+
+use std::{
+    io::ErrorKind,
+    path::{Path, PathBuf},
+};
+
+use vello::wgpu::{self, AdapterInfo, Device, Features, PipelineCache};
+use winit::event_loop::EventLoop;
+
+#[cfg(target_os = "android")]
+fn get_cache_directory_android<T>(event_loop: &EventLoop<T>) -> anyhow::Result<PathBuf> {
+    use std::path::PathBuf;
+
+    use anyhow::Context;
+    use winit::platform::android::EventLoopExtAndroid;
+
+    let app = event_loop.android_app();
+    let app_jobject = unsafe { jni::objects::JObject::from_raw(app.activity_as_ptr().cast()) };
+    // If we got a null VM, we can't pass up
+    let jvm = unsafe { jni::JavaVM::from_raw(app.vm_as_ptr().cast()).context("Making VM")? };
+    let mut env = jvm.attach_current_thread().context("Attaching to thread")?;
+    let res = env
+        .call_method(app_jobject, "getCacheDir", "()Ljava/io/File;", &[])
+        .context("Calling GetCacheDir")?;
+    let file = res.l().context("Converting to JObject")?;
+    let directory_path = env
+        .call_method(file, "getAbsolutePath", "()Ljava/lang/String;", &[])
+        .context("Calling `getAbsolutePath`")?;
+    let string = directory_path.l().context("Converting to a string")?.into();
+    let string = env
+        .get_string(&string)
+        .context("Converting into a Rust string")?;
+    let string: String = string.into();
+    let dir = PathBuf::from(string).join("vello");
+    if !dir.exists() {
+        std::fs::create_dir(&dir).context("Creating pipeline cache directory")?;
+    }
+    // TODO: Also get the quota. This appears to be more involved, requiring a worker thread and being asynchronous
+    Ok(dir)
+}
+
+pub(crate) fn get_cache_directory<T>(
+    _event_loop: &EventLoop<T>,
+) -> anyhow::Result<Option<PathBuf>> {
+    #[cfg(target_os = "android")]
+    {
+        return get_cache_directory_android(_event_loop).map(Some);
+    }
+    #[expect(
+        clippy::allow_attributes,
+        reason = "Doesn't apply if a platform has a cache dir"
+    )]
+    #[allow(
+        unreachable_code,
+        reason = "Fallback path if none of the short-circuits apply"
+    )]
+    return Ok(None);
+}
+
+/// Load a pipeline cache for the given device, or create one if not present.
+///
+/// # Safety
+///
+/// The directory should only have been written into by [`write_pipeline_cache`].
+///
+/// We recognise that this is an impossible standard, but that's because Vulkan
+/// makes UB unavoidable around invalid pipeline caches, so just do your best.
```

Comment:
That's unfortunate. Can we at least embed a checksum in the cache file? If possible, one with a program-local salt so it's extra hard to get the wrong cache by accident?

Reply:
I struggle to see the point. wgpu already does some sanity checks here, and it's still possible to do UB (if *something* decides to write to the wrong place with a forged cache with our hash in place).

I don't believe adding it here for this example to be a good use of my time.

If you think this would be a good use of your time, there is space reserved for a hash in the header wgpu writes. The place to contribute that hashing is: https://github.com/gfx-rs/wgpu/blob/trunk/wgpu-core/src/pipeline_cache.rs. The thing that you'd need to use as salt is probably instead (most of) the data in `PipelineCacheHeader`